modules/shortcodes/instagram.php

```diff
@@ -132,6 +132,79 @@ function jetpack_instagram_embed_reversal( $content ) {
 	return $content;
 }
 
+/**
+ * List of whitelisted and sanitized parameters
+ * that can be used with the Instagram oEmbed endpoint.
+ *
+ * Those parameters can be provided via the Instagram URL, or via shortcode parameters.
+ *
+ * @see https://developers.facebook.com/docs/graph-api/reference/instagram-oembed#parameters
+ *
+ * @since 9.1.0
+ *
+ * @param string $url  URL of the content to be embedded.
+ * @param array  $atts Shortcode attributes.
+ *
+ * @return array $params Array of parameters to be used in Instagram query.
+ */
+function jetpack_instagram_get_whitelisted_parameters( $url, $atts = array() ) {
+	global $content_width;
+
+	// Any URL passed via a shortcode attribute takes precedence.
+	if ( ! empty( $atts['url'] ) ) {
+		$url = $atts['url'];
+		unset( $atts['url'] );
+	}
+
+	/*
+	 * Get URL and parameters from the URL if possible.
+	 *
+	 * We'll also clean any other query params from the URL since Facebook's new API for Instagram
+	 * embeds does not like query parameters. See p7H4VZ-2DU-p2.
+	 */
+	$parsed_url = wp_parse_url( $url );
+	if ( $parsed_url ) {
+		// Bail early if this is not an Instagram URL.
+		if ( ! preg_match( '/instagr(\.am|am\.com)/', $parsed_url['host'] ) ) {
```

Comment:
I'm not particularly happy with having to parse the string multiple times (including once again  in `jetpack_shortcode_instagram` for AMP views); maybe it would be better to use a single regex  to extract everything at once. I'm afraid this would be more fragile though.

Reply:
FWIW, a number of oEmbed related filters use `$url` as a param, and I matched against a RegEx a couple of times (mostly to bail early if the filter was only meant for IG). I did end up using the same RegEx for all of them, which is also the one that the oEmbed handler is registered for, which I think should make it quite robust. (I should've probably extracted the RegEx into a constant at the time, but the cleanest way to do that seemed to be by encapsulating everything into static methods of a class, and that seemed like it'd inflate #17356 even further.)

Reply:
Sorry if I wasn't clear enough. I'm not concerned about this specific regex; I was mostly wondering if we were trying to parse that one URL string too many times:

- We first use `wp_parse_url` above.
- We then `preg_match` here.
- We have another `preg_match` if you're on an AMP view (in `jetpack_shortcode_instagram`). 

I was wondering if it may be worth trying to simplify things a bit by having a single regex we would use to:
- Detect the proper URL and its components
- Bail if it's not an Instagram URL
- Pull out the Instagram unique ID, add it as a shortcode attribute so it can be used by AMP here:
https://github.com/Automattic/jetpack/blob/update/instagram-supported-params/modules/shortcodes/instagram.php#L328

Something a bit like this:

```diff
diff --git a/modules/shortcodes/instagram.php b/modules/shortcodes/instagram.php
index bc4189319..a5b47bccd 100644
--- a/modules/shortcodes/instagram.php
+++ b/modules/shortcodes/instagram.php
@@ -162,23 +162,21 @@ function jetpack_instagram_get_allowed_parameters( $url, $atts = array() ) {
 	 * We'll also clean any other query params from the URL since Facebook's new API for Instagram
 	 * embeds does not like query parameters. See p7H4VZ-2DU-p2.
 	 */
-	$parsed_url = wp_parse_url( $url );
-	if ( $parsed_url && isset( $parsed_url['host'] ) && isset( $parsed_url['path'] ) ) {
-		// Bail early if this is not an Instagram URL.
-		if ( ! preg_match( '/(?:^|\.)instagr(?:\.am|am\.com)$/', $parsed_url['host'] ) ) {
-			return array();
-		}
-
-		$url = 'https://www.instagram.com' . $parsed_url['path'];
+	$parsed_url = array();
+	preg_match( '#https?://(?:www\.|)instagr(?:\.am|am\.com)(/(?:p|tv)/([^/?\#\s]*))?(?:\?([^\#\s]*))?#i', $url, $parsed_url );
+	if ( ! empty( $parsed_url ) && ! empty( $parsed_url[1] ) ) {
+		$url = 'https://www.instagram.com' . $parsed_url[1];
+		$atts['uid'] = $parsed_url[2];
 
 		// If we have any parameters as part of the URL, we merge them with our attributes.
-		if ( ! empty( $parsed_url['query'] ) ) {
+		if ( ! empty( $parsed_url[3] ) ) {
 			$query_args = array();
-			wp_parse_str( $parsed_url['query'], $query_args );
+			wp_parse_str( $parsed_url[3], $query_args );
 
 			$atts = array_merge( $atts, $query_args );
 		}
 	} else {
+		// Bail if this is not an Instagram URL, or not a properly formatted one.
 		return array();
 	}
 
@@ -191,6 +189,7 @@ function jetpack_instagram_get_allowed_parameters( $url, $atts = array() ) {
 			'width'       => isset( $content_width ) ? $content_width : $max_width,
 			'height'      => '',
 			'hidecaption' => false,
+			'uid'         => '',
 		),
 		$atts,
 		'instagram'
@@ -318,14 +317,14 @@ function jetpack_shortcode_instagram( $atts ) {
 	}
 
 	if ( class_exists( 'Jetpack_AMP_Support' ) && Jetpack_AMP_Support::is_amp_request() ) {
-		$url_pattern = '#http(s?)://(www\.)?instagr(\.am|am\.com)/p/([^/?]+)#i';
-		preg_match( $url_pattern, $atts['url'], $matches );
-		$shortcode_id = end( $matches );
+		if ( empty( $atts['uid'] ) ) {
+			return '';
+		}
 		$width        = ! empty( $atts['width'] ) ? $atts['width'] : 600;
 		$height       = ! empty( $atts['height'] ) ? $atts['height'] : 600;
 		return sprintf(
 			'<amp-instagram data-shortcode="%1$s" layout="responsive" width="%2$d" height="%3$d" data-captioned></amp-instagram>',
-			esc_attr( $shortcode_id ),
+			esc_attr( $atts['uid'] ),
 			absint( $width ),
 			absint( $height )
 		);

```

That said, it may be a bit more fragile, and I don't know if parsing 3 times like we do right now is something we should worry about.

cc @anomiex What's your take on this?

Reply:
The regex needs some work, but nothing too major

* It needs anchoring at the start, to avoid matching something like "http://evil.example.com/https://instagram.com/p/whatever".
* It'll fail to match the URLs suggested in the testing instructions, as it doesn't account for path components after the ID.
* If you're only going to accept URLs that have a path component, there's no point in making it optional in the regex. Just let the regex fail to match if it lacks a valid path.
* Same for the ID path component, if you're going to reject a URL with an empty ID you may as well not have the regex accept it in the first place (i.e. use `+` rather than `*` there).
* [nitpick] You could also avoid having to escape the `#` by using a different delimiter.

```php
preg_match( '!https?://(?:www\.|)instagr(?:\.am|am\.com)(/(?:p|tv)/([^/?#\s]+)(?:/[^?#\s]*)?)(?:\?([^#\s]*))?!i', $url, $parsed_url );
```

Are there any other subdomains or paths that might be used by Instagram now or in the future that we'd care about (and not need any code changes other than updating the regex for)?

Also, what happens if someone tries to change the url or uid in the shortcode without updating the other to match? The AMP would use the uid attribute, while normal views would use the url?

Reply:
> That said, it may be a bit more fragile, and I don't know if parsing 3 times like we do right now is something we should worry about.

Inclined to say it's not worth the added fragility :grimacing:

Reply:
> Inclined to say it's not worth the added fragility

Yes that was my worry as well.

> what happens if someone tries to change the url or uid in the shortcode without updating the other to match? The AMP would use the uid attribute, while normal views would use the url?

Yes, that's something I hadn't thought about, you're right.

Overall, this is adding some complexity that may not be needed here. I won't commit this, and I'll add back the check we had discussed here:
https://github.com/Automattic/jetpack/pull/17700#discussion_r516934139